### PR TITLE
Add C#9 tests for S3902,S3885,S1147,S1215,S3971,S1696,S2221,S3981,S2365,S4015

### PR DESCRIPTION
**Pull request by andrei-epure-sonarsource**
Related to #3668

### CODE REVIEW

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/DoNotDecreaseMemberVisibility.CSharp9.cs [hunk 1]
```diff
@@ -0,0 +1,179 @@
+﻿using System;
```
Comment (andrei-epure-sonarsource) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/DoNotDecreaseMemberVisibility.CSharp9.cs:1]: this is the existing test for classes where I replaced:
- `class` to `record`
- `set` to `init`

it's a bit overkill as records are just classes, but it shouldn't hurt.

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/DoNotCopyArraysInProperties.CSharp9.cs [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+record R
+{
+    private static string[] staticStrings = new string[] { "a", "b", "c" };
+    private string[] stringArray = new string[10];
+    public string[] Property1
+    {
+        get { return (string[])stringArray.Clone(); } // Noncompliant {{Refactor 'Property1' into a method, properties should not copy collections.}}
+
+        init { stringArray = (string[])staticStrings.Clone(); }
+    }
```
Comment (pavel-mikula-sonarsource) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/DoNotCopyArraysInProperties.CSharp9.cs:14]: What about this?
```
init { stringArray = value.ToArray(); }
```

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/DoNotDecreaseMemberVisibility.CSharp9.cs [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+
+namespace SomeNamespace
+{
+    public record Class3 : OtherNamespace.Class1
+    {
+        private void SomeMethod(string s) { }
+    }
+}
+
+namespace FalsePositiveOnIndexers
```
Comment (pavel-mikula-sonarsource) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/DoNotDecreaseMemberVisibility.CSharp9.cs:161]: It's not a FP here
```suggestion
namespace Indexers
```

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
Few improvements suggested

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/DoNotDecreaseMemberVisibility.CSharp9.cs (hunk 1 above)
Comment (pavel-mikula-sonarsource) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/DoNotDecreaseMemberVisibility.CSharp9.cs:1]: It's `Implement`, so it's useful like this

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/Rules/DoNotCallGCSuppressFinalizeTest.cs [hunk 4]
```diff
@@ -31,13 +32,18 @@ public class DoNotCallGCSuppressFinalizeTest
...
         [TestMethod]
         [TestCategory("Rule")]
         public void DoNotCallGCSuppressFinalize_NetCore() =>
-            Verifier.VerifyAnalyzer(@"TestCases\DoNotCallGCSuppressFinalize.NetCore.cs", new DoNotCallGCSuppressFinalizeMethod(), ParseOptionsHelper.FromCSharp8);
+            Verifier.VerifyAnalyzer(@"TestCases\DoNotCallGCSuppressFinalize.NetCore.cs", new DoNotCallGCSuppressFinalize(), ParseOptionsHelper.FromCSharp8);
+
+        [TestMethod]
+        [TestCategory("Rule")]
+        public void DoNotCallGCSuppressFinalize_Net() =>
+            Verifier.VerifyAnalyzer(@"TestCases\DoNotCallGCSuppressFinalize.Net.cs", new DoNotCallGCSuppressFinalize(), ParseOptionsHelper.FromCSharp9);
```
[the remark this answers is not in the document]
Comment (andrei-epure-sonarsource) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/Rules/DoNotCallGCSuppressFinalizeTest.cs:46]: It's not NetCore, it's .NET

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/DoNotCheckZeroSizeCollection.CSharp9.cs [hunk 5]
```diff
@@ -0,0 +1,48 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+
+var someEnumerable = new List<string>();
+
+var result = someEnumerable.Count() >= 0; // Noncompliant {{The count of 'IEnumerable<T>' is always '>=0', so fix this test to get the real expected behavior.}}
+
+if (someEnumerable.Count() is >= 0) { } // FN
```
[the remark this answers is not in the document]
Comment (andrei-epure-sonarsource) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/DoNotCheckZeroSizeCollection.CSharp9.cs:9]: it is for `Pattern matching improvements`

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/Rules/DoNotCallGCSuppressFinalizeTest.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/Rules/DoNotCallGCSuppressFinalizeTest.cs:46]: It's under `#if NETCOREAPP` section, so it's confusing. Can you use `DoNotCallGCSuppressFinalize.Net5.cs` then?

**Review by pavel-mikula-sonarsource — APPROVED** [review]
LGTM. Please rebase and use the new methods before merging.

**Pull request merged**